src/global/TransitionManager.ts

```diff
@@ -0,0 +1,179 @@
+    this.check();
+  }
+
+  public start(): void {
+    this.started = true;
+    this.children.forEach(c => c.start());
+    this.intros.concat(this.outros).forEach(t => t.start());
+    this.check();
+  }
+
+  private check(): void {
+    if (!this.started || this.outros.length || this.outroChildren) return;
+
+    // If all outros are complete, and we haven't already done this,
+    // we notify the parent if there is one, otherwise
+    // start detaching nodes
+    if (!this.outrosComplete) {
+      this.outrosComplete = true;
+
+      if (this.parent) this.parent.decrementOutros();
+
+      if (TransitionManager.allOutrosComplete(this)) {
+        this.detachNodes();
+      }
+    }
+
+    // Once everything is done, we can notify parent transition
+    // manager and call the callback
+    if (!this.intros.length && !this.totalChildren) {
+      if (isFunction(this.callback)) {
+        this.callback();
+      }
+
+      if (this.parent && !this.notifiedTotal) {
+        this.notifiedTotal = true;
+        this.parent.decrementTotal();
+      }
+    }
+  }
+
+  // check through the detach queue to see if a node is up or downstream from a
+  // transition and if not, go ahead and detach it
+  private detachImmediate(): void {
+    const queue = this.detachQueue;
+    const outros = TransitionManager.collectAllOutros(this);
+
+    if (!outros.length) {
+      this.detachNodes();
+    } else {
+      let i = queue.length;
+      let j = 0;
+      let node, trans;
+      const nqueue = (this.detachQueue = []);
+
+      start: while (i--) {
+        node = queue[i].node;
+        j = outros.length;
+        while (j--) {
+          trans = outros[j].element.node;
+          // check to see if the node is, contains, or is contained by the transitioning node
+          if (trans === node || trans.contains(node) || node.contains(trans)) {
+            nqueue.push(queue[i]);
+            continue start;
+          }
+        }
+
+        // no match, we can drop it
+        queue[i].detach();
+      }
+    }
+  }
+
+  private static allOutrosComplete(manager: TransitionManager): boolean {
+    return (
+      !manager || (manager.outrosComplete && TransitionManager.allOutrosComplete(manager.parent))
+    );
+  }
+
+  private static collectAllOutros(manager: TransitionManager, _list?): any[] {
+    let list = _list;
+
+    // if there's no list, we're starting at the root to build one
+    if (!list) {
+      list = [];
+      let parent = manager;
+      while (parent.parent) parent = parent.parent;
+      return TransitionManager.collectAllOutros(parent, list);
+    } else {
+      // grab all outros from child managers
+      let i = manager.children.length;
+      while (i--) {
+        list = TransitionManager.collectAllOutros(manager.children[i], list);
+      }
+
+      // grab any from this manager if there are any
+      if (manager.outros.length) list = list.concat(manager.outros);
+
+      return list;
+    }
+  }
```

Comment:
This is an interesting scenario.
We have converted a non-exported module function into a private static method.

From one point of view, it makes sense to convert `collectAllOutros` into a private function, however we noticed that TypeScript makes all private functions publicly available after they’re compiled.

The big difference here would be that you could technically call `TransitionManager.collectAllOutros(....)` from another JS file.

@evs-chris do you think that it would be a problem?   I’d say we pick one approach and ideally stick to it for every similar scenario in Ractive codebase.

Reply:
I lean more towards module functions than methods because they minify and tree shake so much better. In fact, part of my long term plan is to reduce some of the classiness in the code base, as it makes things a bit heavier than they should be.

Reply:
Ok, I'll revert the static methods added in Runloop and TransitionManagar